sky/cloud_stores.py

```diff
@@ -153,6 +155,113 @@ def make_sync_file_command(self, source: str, destination: str) -> str:
+    # Hence, az-cli for SAS token generation is ran on the local machine and
+    # AzCopy is installed at the remote machine for efficient data transfer
+    # from containers to remote systems.
+    # Note that on Azure instances, both az-cli and AzCopy are typically
+    # pre-installed. And installing both would be used with AZ container is
+    # used from non-Azure instances.
+
+    _GET_AZCOPY = [
+        'azcopy --version > /dev/null 2>&1 || '
+        '(mkdir -p /usr/local/bin; '
+        'curl -L https://aka.ms/downloadazcopy-v10-linux -o azcopy.tar.gz; '
+        'sudo tar -xvzf azcopy.tar.gz --strip-components=1 -C /usr/local/bin --exclude=*.txt; '  # pylint: disable=line-too-long
+        'sudo chmod +x /usr/local/bin/azcopy; '
+        'rm azcopy.tar.gz)'
+    ]
+
+    def is_directory(self, url: str) -> bool:
+        """Returns whether 'url' of the AZ Container is a directory.
+
+        In cloud object stores, a "directory" refers to a regular object whose
+        name is a prefix of other objects.
+        """
+        # split the url using split_az_path
+        _, _, path = data_utils.split_az_path(url)
+        # If there aren't more than just container name and storage account,
+        # that's a directory.
+        if not path:
+            return True
+        # If there's more, we'd need to check if it's a directory or a file.
+        container_client = data_utils.create_az_client(client_type='container',
+                                                       container_url=url)
+        num_objects = 0
+        for blob in container_client.list_blobs(name_starts_with=path):
+            if blob.name == path:
+                return False
+            num_objects += 1
+            if num_objects > 1:
+                return True
+
+        # A directory with few or no items
+        return True
+
+    def _get_azcopy_source(self, source: str, is_dir: bool) -> str:
+        """Converts the source so it can be used as an argument for azcopy."""
+        storage_account_name, container_name, blob_path = (
+            data_utils.split_az_path(source))
+        resource_group_name = data_utils.get_az_resource_group(
+            storage_account_name)
+        storage_account_key = data_utils.get_az_storage_account_key(
+            storage_account_name, resource_group_name)
+
+        if storage_account_key is None:
+            # public containers do not require SAS token for access
+            sas_token = ''
+        else:
+            if is_dir:
+                sas_token = azure.get_az_container_sas_token(
+                    storage_account_name, storage_account_key, container_name)
+            else:
+                sas_token = azure.get_az_blob_sas_token(storage_account_name,
+                                                        storage_account_key,
+                                                        container_name,
+                                                        blob_path)
```

Comment:
```suggestion
        if storage_account_key is None:
            # public containers do not require SAS token for access
            return shlex.quote(f'{source}/')
        if is_dir:
            sas_token = azure.get_az_container_sas_token(
                storage_account_name, storage_account_key, container_name)
        else:
            sas_token = azure.get_az_blob_sas_token(storage_account_name,
                                                        storage_account_key,
                                                        container_name,
                                                        blob_path)
```
nit; to reduce code indention